# Wikipedia Notable Life Expectancies

# [Notebook 2 of 4: Data Cleaning](https://github.com/teresahanak/wikipedia-life-expectancy/blob/main/wp_life_expect_data_clean_thanak_2022_06_13.ipynb)

## Context

The


## Objective

The

### Data Dictionary

- Feature: Description

## Importing Necessary Libraries

In [1]:
# To structure code automatically
%load_ext nb_black

# To import/export sqlite databases
import sqlite3 as sql

# To help with reading and manipulating data
import pandas as pd
import numpy as np
import re

# To define maximum number of columns to be displayed in a dataframe
pd.set_option("display.max_columns", None)
# To define the maximum number of rows to be displayed in a dataframe
pd.set_option("display.max_rows", 200)

# To supress warnings
# import warnings

# warnings.filterwarnings("ignore")

# To set some visualization attributes
pd.set_option("max_colwidth", 150)

<IPython.core.display.Javascript object>

## Data Overview

### Reading, Sampling, and Checking Data Shape

In [2]:
# Reading the wp_life_expect_raw_complete dataset
conn = sql.connect("wp_life_expect_raw_complete.db")
data = pd.read_sql("SELECT * FROM wp_life_expect_raw_complete", conn)

# Making a working copy
df = data.copy()

# Checking the shape
print(f"There are {df.shape[0]} rows and {df.shape[1]} columns.")

# Checking first 2 rows of the data
df.head(2)

There are 133900 rows and 6 columns.


month_year day              name  \
0  January 1994   1  William Chappell   
1  January 1994   1    Raymond Crotty   

                                                  info  \
0  , 86, British dancer, ballet designer and director.   
1         , 68, Irish economist, writer, and academic.   

                                                      link num_references  
0  https://en.wikipedia.org/wiki/William_Chappell_(dancer)             21  
1             https://en.wikipedia.org/wiki/Raymond_Crotty             12

<IPython.core.display.Javascript object>

In [3]:
# Checking last 2 rows of the data
df.tail(2)

month_year day           name  \
133898  June 2022   9  Oleg Moliboga   
133899  June 2022   9       Zou Jing   

                                                                         info  \
133898    , 69, Russian volleyball player, Olympic champion (1980) and coach.   
133899  , 86, Chinese engineer, member of the Chinese Academy of Engineering.   

                                                     link num_references  
133898        https://en.wikipedia.org/wiki/Oleg_Moliboga              2  
133899  https://en.wikipedia.org/wiki/Zou_Jing_(engineer)              3

<IPython.core.display.Javascript object>

In [4]:
# Checking a sample of the data
df.sample(5)

month_year day                                     name  \
111904     April 2020  14                                Akin Euba   
89727        May 2017   7  Hugh Thomas, Baron Thomas of Swynnerton   
114983    August 2020   2                              Larry Novak   
13654        May 1999   1                                Jos LeDuc   
108180  November 2019  30                            Milagrosa Tan   

                                                                                         info  \
111904                                                               , 84, Nigerian musician.   
89727                                                  , 85, British historian and life peer.   
114983                                                           , 87, American jazz pianist.   
13654                                   , 54, Canadian professional wrestler, lung infection.   
108180  , 61, Filipino politician, Governor of Samar (2001–2010, since 2019), cardiac arrest.   

                                                                         link  \
111904                                https://en.wikipedia.org/wiki/Akin_Euba   
89727   https://en.wikipedia.org/wiki/Hugh_Thomas,_Baron_Thomas_of_Swynnerton   
114983                              https://en.wikipedia.org/wiki/Larry_Novak   
13654                                 https://en.wikipedia.org/wiki/Jos_LeDuc   
108180                            https://en.wikipedia.org/wiki/Milagrosa_Tan   

       num_references  
111904              1  
89727              11  
114983              2  
13654              44  
108180             10

<IPython.core.display.Javascript object>

#### Observations:
- There are 133,900 rows and 6 columns.
- `month_year` contains the month and year of death, while `day` contains the day of the month of death.
- `name` is the notable person's name.  It is a nominal feature that will not be used for analysis, but will be maintained for any referencing needs.
- `info` contains multiple items including the notable person's "age, country of citizenship at birth, subsequent country of citizenship (if applicable), reason for notability, (and) cause of death (if known)."
- `link` is the url to the notable person's individual Wikipedia page.  If such a page does not exist, there is either a non-working link (https://en.wikipedia.orgNone), or the link is to a page with a message that the page does not exist for that individual.  `link` is a unique identifier for all entries, except the 6 with the non-working link, which do have unique `name` values from each other.
- `num_references` contains the number of references on the notable person's individual Wikipedia page.  This feature serves as a proxy measure of notability.
- Prior to EDA, our task will be to extract the individual elements that are comined in `month_year` and `info` columns.

### Checking Data Types, Duplicates, and Null Values

In [5]:
# Checking data types and null values
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 133900 entries, 0 to 133899
Data columns (total 6 columns):
 #   Column          Non-Null Count   Dtype 
---  ------          --------------   ----- 
 0   month_year      133900 non-null  object
 1   day             133900 non-null  object
 2   name            133894 non-null  object
 3   info            133900 non-null  object
 4   link            133900 non-null  object
 5   num_references  133900 non-null  object
dtypes: object(6)
memory usage: 6.1+ MB


<IPython.core.display.Javascript object>

In [6]:
# Checking duplicate rows
df.duplicated().sum()

0

<IPython.core.display.Javascript object>

In [7]:
# Check percentage of null values by column
df.isnull().sum() / df.count() * 100

month_year        0.000000
day               0.000000
name              0.004481
info              0.000000
link              0.000000
num_references    0.000000
dtype: float64

<IPython.core.display.Javascript object>

In [8]:
# Checking number of missing values per row (not necessary here, but done to keep process standard)
df.isnull().sum(axis=1).value_counts()

0    133894
1         6
dtype: int64

<IPython.core.display.Javascript object>

#### Observations:
- Our dataset was saved to and read from the database without any hiccups.
- As expected, we have 6 entries that are missing `name`, but we will find it in their `info` values.
- All columns are currently of object type.  We will need to appropriately typecast them after separating the information in `month_year` and `info`.

## Data Cleaning

### Addressing Missing `name` Values

In [9]:
# Checking rows with missing name values
missing_name = df[df["name"].isna()]
missing_name

month_year day  name  \
18937      August 2001  11  None   
24985     January 2004  22  None   
27458       March 2005   1  None   
34077        July 2007  11  None   
64769   September 2013  29  None   
106613  September 2019  12  None   

                                                                                      info  \
18937   Kevin Kowalcyk, 2, known for eating a hamburger contaminated with E. coli O157:H7.   
24985                                                Vincent Palmer, 37, British criminal.   
27458                                             Barry Stigler, 57, American voice actor.   
34077                                          Nana Gualdi, 75, German singer and actress.   
64769                                 Scott Workman, 47, American stuntman (, , ), cancer.   
106613                                                Thami Shobede, 31, Singer Songwriter   

                                link num_references  
18937   https://en.wikipedia.orgNone              0  
24985   https://en.wikipedia.orgNone              0  
27458   https://en.wikipedia.orgNone              0  
34077   https://en.wikipedia.orgNone              0  
64769   https://en.wikipedia.orgNone              0  
106613  https://en.wikipedia.orgNone              0

<IPython.core.display.Javascript object>

#### Observations:
- These rows vary from the main set as there is a substring containing the person's name at the start of the `info` string.
- As there are so few rows missing `name`, let us address this issue first.

In [10]:
# For loop to copy name value from info value and remove name from info value
treat_rows = missing_name.index
for i in treat_rows:
    info = df.loc[i, "info"]
    info_lst = info.split(sep=",", maxsplit=1)

    name = info_lst[0].strip()
    df.loc[i, "name"] = name
    df.loc[i, "info"] = re.sub(name, "", info).strip()

# Re-check rows
df.loc[treat_rows, :]

month_year day            name  \
18937      August 2001  11  Kevin Kowalcyk   
24985     January 2004  22  Vincent Palmer   
27458       March 2005   1   Barry Stigler   
34077        July 2007  11     Nana Gualdi   
64769   September 2013  29   Scott Workman   
106613  September 2019  12   Thami Shobede   

                                                                        info  \
18937   , 2, known for eating a hamburger contaminated with E. coli O157:H7.   
24985                                                , 37, British criminal.   
27458                                            , 57, American voice actor.   
34077                                       , 75, German singer and actress.   
64769                                , 47, American stuntman (, , ), cancer.   
106613                                               , 31, Singer Songwriter   

                                link num_references  
18937   https://en.wikipedia.orgNone              0  
24985   https://en.wikipedia.orgNone              0  
27458   https://en.wikipedia.orgNone              0  
34077   https://en.wikipedia.orgNone              0  
64769   https://en.wikipedia.orgNone              0  
106613  https://en.wikipedia.orgNone              0

<IPython.core.display.Javascript object>

#### Observations:
- Missing `name` values have been addressed and those names have been removed from `info` values.

In [11]:
# Re-check info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 133900 entries, 0 to 133899
Data columns (total 6 columns):
 #   Column          Non-Null Count   Dtype 
---  ------          --------------   ----- 
 0   month_year      133900 non-null  object
 1   day             133900 non-null  object
 2   name            133900 non-null  object
 3   info            133900 non-null  object
 4   link            133900 non-null  object
 5   num_references  133900 non-null  object
dtypes: object(6)
memory usage: 6.1+ MB


<IPython.core.display.Javascript object>

#### Observations:
- We have no remaining missing values.
- Let us treat the `month_year` column next.

### Separating `month` and `year`

In [12]:
# Separating month and year into 2 columns and typecasting year as integer
df.loc[:, "year"] = df["month_year"].apply(lambda x: x.split(sep=" ")[1].strip())
df["year"] = df["year"].apply(lambda x: int(x))

df.loc[:, "month"] = df["month_year"].apply(lambda x: x.split(sep=" ")[0])
df.head(2)

month_year day              name  \
0  January 1994   1  William Chappell   
1  January 1994   1    Raymond Crotty   

                                                  info  \
0  , 86, British dancer, ballet designer and director.   
1         , 68, Irish economist, writer, and academic.   

                                                      link num_references  \
0  https://en.wikipedia.org/wiki/William_Chappell_(dancer)             21   
1             https://en.wikipedia.org/wiki/Raymond_Crotty             12   

   year    month  
0  1994  January  
1  1994  January

<IPython.core.display.Javascript object>

In [13]:
# Dropping month_year column
df.drop("month_year", axis=1, inplace=True)
df.head(2)

day              name                                                 info  \
0   1  William Chappell  , 86, British dancer, ballet designer and director.   
1   1    Raymond Crotty         , 68, Irish economist, writer, and academic.   

                                                      link num_references  \
0  https://en.wikipedia.org/wiki/William_Chappell_(dancer)             21   
1             https://en.wikipedia.org/wiki/Raymond_Crotty             12   

   year    month  
0  1994  January  
1  1994  January

<IPython.core.display.Javascript object>

### Treating `info`

#### Checking a Sample

In [14]:
# Checking a sample of info
df.sample(5)

day                 name  \
74610  16  Alexander Melentyev   
30253  19         Mohammad Ali   
16365   8      S. Nijalingappa   
11630   7           Wally Gold   
9781   10       George Malcolm   

                                                                  info  \
74610     , 60, Soviet Russian sport shooter, Olympic champion (1980).   
30253                           , 78, Pakistani actor, cardiac arrest.   
16365                                         , 97, Indian politician.   
11630   , 70, American musician and music business executive, colitis.   
9781   , 80, English pianist, harpsichordist, composer, and conductor.   

                                                          link num_references  \
74610        https://en.wikipedia.org/wiki/Alexander_Melentyev              1   
30253       https://en.wikipedia.org/wiki/Mohammad_Ali_(actor)             11   
16365            https://en.wikipedia.org/wiki/S._Nijalingappa             11   
11630                 https://en.wikipedia.org/wiki/Wally_Gold              5   
9781   https://en.wikipedia.org/wiki/George_Malcolm_(musician)             10   

       year     month  
74610  2015  February  
30253  2006     March  
16365  2000    August  
11630  1998      June  
9781   1997   October

<IPython.core.display.Javascript object>

#### Observations:
- First, let us check for any rows that are missing digits, and therefore the age target, within `info` and remove them.
- Also, it would be helpful to remove information contained within parentheses, as we will not be using it.

#### Checking and Dropping Rows Lacking Digits (and therefore Age Data) within `info`

In [15]:
# For loop to extract index of rows without digits in info value
remove_lst = []
for index in df.index:
    pattern = r"\d"
    if re.search(pattern, df.loc[index, "info"]) is None:
        remove_lst.append(index)
print(len(remove_lst), "rows")
df.loc[remove_lst, :].sample(5)

925 rows


day                 name  \
61552    3  Cornelius the First   
76424   23          Zedi Feruzi   
17055   30        Ansumane Mané   
112266  24        Mustafa Koçak   
57287   24             Dadullah   

                                                          info  \
61552     ,  black rhinoceros, Canadian politician, pneumonia.   
76424                            , Burundian politician, shot.   
17055        , Bissau-Guinean soldier and rebel, assassinated.   
112266  , Turkish political prisoner, self-imposed starvation.   
57287   , Pakistani Taliban leader (Bajaur Agency), airstrike.   

                                                              link  \
61552            https://en.wikipedia.org/wiki/Cornelius_the_First   
76424                    https://en.wikipedia.org/wiki/Zedi_Feruzi   
17055             https://en.wikipedia.org/wiki/Ansumane_Man%C3%A9   
112266            https://en.wikipedia.org/wiki/Mustafa_Ko%C3%A7ak   
57287   https://en.wikipedia.org/wiki/Dadullah_(Pakistani_Taliban)   

       num_references  year     month  
61552               5  2013     April  
76424               7  2015       May  
17055               8  2000  November  
112266              4  2020     April  
57287               9  2012    August

<IPython.core.display.Javascript object>

In [16]:
# Dropping rows missing age data, resetting index, and checking new shape of df
df.drop(remove_lst, inplace=True)
df.reset_index(inplace=True, drop=True)
df.shape

(132975, 7)

<IPython.core.display.Javascript object>

#### Observations:
- 925 rows were removed as they lacked the target data for `age`.

#### Removing Information within Parentheses in `info`

In [17]:
# Regular expression for parenthesis and its contents
pattern = r"\(.*\)"

# Subbing empty string for parentheses and stripping white space
df.loc[:, "info"] = df["info"].apply(lambda x: re.sub(pattern, "", x).strip())
df.loc[[21850, 125430], :]

day                 name  \
21850   11  Sir Michael Clapham   
125430  23       Jean-Luc Nancy   

                                                                                                      info  \
21850   , 90, British industrialist, president of the Confederation of British Industry from 1972 to 1974.   
125430                                                                           , 81, French philosopher.   

                                                                 link  \
21850   https://en.wikipedia.org/wiki/Michael_Clapham_(industrialist)   
125430                   https://en.wikipedia.org/wiki/Jean-Luc_Nancy   

       num_references  year     month  
21850              14  2002  November  
125430              8  2021    August

<IPython.core.display.Javascript object>

#### Observation:
- Parentheses and information within has been removed from `info`.
- Next, we will follow the Wikipedia-defined fields to divide the `info` values.

#### Splitting `info` on Commas into Separate Columns

In [18]:
# For loop to split info on commas and separate into respective new columns and removing leading/trailing white space and periods
for i, item in enumerate(df["info"]):
    info_lst = item.split(",")

    for j in range(len(info_lst)):
        df.loc[i, f"info_{j}"] = info_lst[j].strip(" .")

# Checking the first 2 rows
df.head(2)

day              name                                                 info  \
0   1  William Chappell  , 86, British dancer, ballet designer and director.   
1   1    Raymond Crotty         , 68, Irish economist, writer, and academic.   

                                                      link num_references  \
0  https://en.wikipedia.org/wiki/William_Chappell_(dancer)             21   
1             https://en.wikipedia.org/wiki/Raymond_Crotty             12   

   year    month info_0 info_1           info_2                        info_3  \
0  1994  January            86   British dancer  ballet designer and director   
1  1994  January            68  Irish economist                        writer   

         info_4 info_5 info_6 info_7 info_8 info_9 info_10 info_11  
0           NaN    NaN    NaN    NaN    NaN    NaN     NaN     NaN  
1  and academic    NaN    NaN    NaN    NaN    NaN     NaN     NaN

<IPython.core.display.Javascript object>

In [19]:
# Checking the last 2 rows
df.tail(2)

day           name  \
132973   9  Oleg Moliboga   
132974   9       Zou Jing   

                                                                         info  \
132973          , 69, Russian volleyball player, Olympic champion  and coach.   
132974  , 86, Chinese engineer, member of the Chinese Academy of Engineering.   

                                                     link num_references  \
132973        https://en.wikipedia.org/wiki/Oleg_Moliboga              2   
132974  https://en.wikipedia.org/wiki/Zou_Jing_(engineer)              3   

        year month info_0 info_1                     info_2  \
132973  2022  June            69  Russian volleyball player   
132974  2022  June            86           Chinese engineer   

                                              info_3 info_4 info_5 info_6  \
132973                   Olympic champion  and coach    NaN    NaN    NaN   
132974  member of the Chinese Academy of Engineering    NaN    NaN    NaN   

       info_7 info_8 info_9 info_10 info_11  
132973    NaN    NaN    NaN     NaN     NaN  
132974    NaN    NaN    NaN     NaN     NaN

<IPython.core.display.Javascript object>

#### Observations:
- The `info` value is successfully divided and we can proceed through it column by column.
- We will check the set of values for the first two columns, for age.

### `info_0`

In [20]:
# Checking unique value counts
df["info_0"].value_counts()

                                                                       132946
Sir                                                                         6
92                                                                          2
Douglas Scott                                                               1
VC                                                                          1
Sir Woolwich West                                                           1
Sir Lord Justice of Appeal                                                  1
Dame MEP                                                                    1
83                                                                          1
Sir Governor-General                                                        1
Notable ice hockey players and coaches among the 44 killed in the :         1
Mike Alexander                                                              1
Colonel                                                         

<IPython.core.display.Javascript object>

#### Observations:
- The vast majority of rows have an empty string for this field.
- There is one row representing a group, rather than an individual, and we will drop it.
- We should verify the name and age information for the remainder of unique values in `info_0`.

#### Dropping Entry for Group

In [21]:
# Checking the entry representing a group
group_entry = df[
    df["info_0"]
    == "Notable ice hockey players and coaches among the 44 killed in the :"
]
group_entry

day                                  name  \
51130   7  2011 Lokomotiv Yaroslavl plane crash   

                                                                      info  \
51130  Notable ice hockey players and coaches among the 44 killed in the :   

                                                                     link  \
51130  https://en.wikipedia.org/wiki/2011_Lokomotiv_Yaroslavl_plane_crash   

      num_references  year      month  \
51130             95  2011  September   

                                                                    info_0  \
51130  Notable ice hockey players and coaches among the 44 killed in the :   

      info_1 info_2 info_3 info_4 info_5 info_6 info_7 info_8 info_9 info_10  \
51130    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN     NaN   

      info_11  
51130     NaN

<IPython.core.display.Javascript object>

In [22]:
# Dropping group entry, resetting index, and checking new shape of df
df.drop(group_entry.index, inplace=True)
df.reset_index(inplace=True, drop=True)
df.shape

(132974, 19)

<IPython.core.display.Javascript object>

#### Examining Rows with Atypical `info_0` Values

In [23]:
# Examining rows with atypical info_0 values
list_to_check = df["info_0"].value_counts().index.to_list()

verify_df = pd.DataFrame()
for item in list_to_check[1:]:
    verify_df = pd.concat([verify_df, df[df["info_0"] == item]])
verify_df

day                     name  \
30349   25           Julian Bullard   
31140    1          Kyffin Williams   
34207   15             Jeremy Moore   
43606   29         Derek Hodgkinson   
67212    7             Richard Best   
67217    7             Thomas Hinde   
33086   28           David Turnbull   
117387  11         Gotthilf Fischer   
67517   21             Colin Turner   
67162    5               Robin Dunn   
67152    4    Elaine Kellett-Bowman   
59616   24      Kristján Jóhannsson   
52580   25           Moses Pitakaka   
42140    5                    Evile   
40355   22          Dudley Thornton   
34171    9       Sir Tasker Watkins   
33751   22             John Burnett   
30468    8           Peter Smithers   
14035    5                slapstick   
30363   26       Mahmoud al-Majzoub   
30358   26  Horondino José da Silva   
30210    6        Flight Lieutenant   
30188    2         Sam Mokuahi, Jr.   
29619   10             James Yancey   
18908   11              were killed   
18312   20            Demetreus Nix   
15873   15                      NRG   
67661   30            David Gibbons   

                                                                                                                 info  \
30349                                                                                     Sir , 78, British diplomat.   
31140                                                               Sir , 88, Welsh artist, lung and prostate cancer.   
34207                                    Sir , 79, British soldier, commander of UK land forces in the Falklands War.   
43606                                                                            Sir , 92, British air chief marshal.   
67212                                                             Sir , 80, British diplomat, Ambassador to Iceland .   
67217                                                                                     Sir , 88, British novelist.   
33086                                                                             . 92, American materials scientist.   
117387                                                                                  92, German choral conductor .   
67517                                                            Sir Woolwich West, 92, British politician, MP for  .   
67162                                                              Sir Lord Justice of Appeal, 96, British jurist,  .   
67152                                              Dame MEP, 90, British politician and barrister, MP for Lancaster .   
59616                                                                                  83, Icelandic Olympic athlete.   
52580                                                       Sir Governor-General, 66, Solomon Islander politician,  .   
42140                                                       Mike Alexander, 32, British bassist , pulmonary embolism.   
40355                                                                             Colonel , 89, British army officer.   
34171   VC, 88, British jurist and businessman, Lord Justice of Appeal and former WRU President, after short illness.   
33751                                                 Sir Edinburgh University, 85, British academic, Principal of  .   
30468                 Sir Winchester, 92, British politician, MP for Secretary General of the Council of Europe and .   
14035                                Herbert Wiere, 91, Austrian-born American Wiere Brothers comedian, member of the   
30363                                  aka Abu Hamza, 41, Palestinian Islamic Jihad leader, assassination by bombing.   
30358                                     aka "Dino Sete Cordas", 88, Brazilian virtuouso of the seven-string guitar.   
30210                           Sarah-Jayne Mulvihill, 32, first British servicewoman to be killed in action in Iraq.   
30188     aka "Sammy Steamboat", 71, American Hawaiian professional wrestler, complications 

<IPython.core.display.Javascript object>

#### Observations:
- The majority of rows contain additional aliases or titles within `info_0`, that we don't need, but we can leave in place for now.  
- There are a few rows that will need to be treated individually to correct the name value, as follows:
    1. Entry is for Mike Alexander whose band was Evile.
    2. Entry is for Herbert Wiere who performed slapstick.
    3. Entry is for Sarah-Jayne Mulvihill who was a Flight Lieutenant.
    4. Entry is for Douglass Scott who was killed by Demetreus Nix.
    5. Entry is for Kim Hwan-Sung who was a member of the band NRG.
- We can replace the `name` value with the `info_0` value for these rows as well as proceed with hard-coding the correct values for info_2 and info_3 fields to match the Wikipedia pattern, but staying true to the information scraped.
- The row with "Nearly 3" value for `info_0` represents a group, rather than an individual, so will be dropped, after treating the above rows.
- We can proceed to extract age from `info_0` for the few rows that contain it here instead of in `info_1`.

#### Treating 5 rows with Name in `info_0`

In [24]:
values_lst = [
    "Mike Alexander",
    "Herbert Wiere",
    "Sarah-Jayne Mulvihill",
    "Douglas Scott",
    "Kim Hwan-Sung",
]

<IPython.core.display.Javascript object>

In [25]:
# For loop to copy name from info_0 to name
for i in df[df["info_0"].isin(values_lst)].index.to_list():
    df.loc[i, "name"] = df.loc[i, "info_0"]

# Hard-coding info_2 and info_3 values for Kim Hwan-Sung
index = df[
    df["link"] == "https://en.wikipedia.org/wiki/NRG_(South_Korean_band)"
].index.to_list()
df.loc[index, "info_2"] = "South Korean musician"

df.loc[index, "info_3"] = "respiratory illness"

# # Hard-coding info_2 and info_3 values for Douglass Scott
index = df[
    df["link"]
    == "https://en.wikipedia.org/w/index.php?title=Demetreus_Nix&action=edit&redlink=1"
].index.to_list()
df.loc[index, "info_2"] = "student"

df.loc[index, "info_3"] = "murdered"

# # Hard-coding info_2 and info_3 values for Sarah-Jayne Mulvihill
index = df[
    df["link"] == "https://en.wikipedia.org/wiki/Flight_Lieutenant"
].index.to_list()
df.loc[index, "info_2"] = "British servicewoman"

df.loc[index, "info_3"] = "killed in action"

<IPython.core.display.Javascript object>

In [26]:
df[df["info_0"].isin(values_lst)]

day                   name  \
14035   5          Herbert Wiere   
15873  15          Kim Hwan-Sung   
18312  20          Douglas Scott   
30210   6  Sarah-Jayne Mulvihill   
42140   5         Mike Alexander   

                                                                                        info  \
14035       Herbert Wiere, 91, Austrian-born American Wiere Brothers comedian, member of the   
15873                                                       Kim Hwan-Sung, 19, A Member of .   
18312                                   Douglas Scott, 20, High-school student murdered by .   
30210  Sarah-Jayne Mulvihill, 32, first British servicewoman to be killed in action in Iraq.   
42140                              Mike Alexander, 32, British bassist , pulmonary embolism.   

                                                                                 link  \
14035                                         https://en.wikipedia.org/wiki/Slapstick   
15873                           https://en.wikipedia.org/wiki/NRG_(South_Korean_band)   
18312  https://en.wikipedia.org/w/index.php?title=Demetreus_Nix&action=edit&redlink=1   
30210                                 https://en.wikipedia.org/wiki/Flight_Lieutenant   
42140                                             https://en.wikipedia.org/wiki/Evile   

      num_references  year    month                 info_0 info_1  \
14035             16  1999   August          Herbert Wiere     91   
15873             27  2000     June          Kim Hwan-Sung     19   
18312              0  2001     June          Douglas Scott     20   
30210             12  2006      May  Sarah-Jayne Mulvihill     32   
42140             82  2009  October         Mike Alexander     32   

                                               info_2               info_3  \
14035  Austrian-born American Wiere Brothers comedian        member of the   
15873                           South Korean musician  respiratory illness   
18312                                         student             murdered   
30210                            British servicewoman     killed in action   
42140                                 British bassist   pulmonary embolism   

      info_4 info_5 info_6 info_7 info_8 info_9 info_10 info_11  
14035    NaN    NaN    NaN    NaN    NaN    NaN     NaN     NaN  
15873    NaN    NaN    NaN    NaN    NaN    NaN     NaN     NaN  
18312    NaN    NaN    NaN    NaN    NaN    NaN     NaN     NaN  
30210    NaN    NaN    NaN    NaN    NaN    NaN     NaN     NaN  
42140    NaN    NaN    NaN    NaN    NaN    NaN     NaN     NaN

<IPython.core.display.Javascript object>

#### Dropping Entry for Group

In [27]:
# Checking the entry representing a group
group_entry = df[df["info_0"] == "Nearly 3"]
group_entry

day         name  \
18908  11  were killed   

                                                               info  \
18908  Nearly 3,000 people September 11 attacks in the , including:   

                                                                       link  \
18908  https://en.wikipedia.org/wiki/Casualties_of_the_September_11_attacks   

      num_references  year      month    info_0  \
18908            176  2001  September  Nearly 3   

                                       info_1      info_2 info_3 info_4  \
18908  000 people September 11 attacks in the  including:    NaN    NaN   

      info_5 info_6 info_7 info_8 info_9 info_10 info_11  
18908    NaN    NaN    NaN    NaN    NaN     NaN     NaN

<IPython.core.display.Javascript object>

In [28]:
# Dropping group entry, resetting index, and checking new shape of df
df.drop(group_entry.index, inplace=True)
df.reset_index(inplace=True, drop=True)
df.shape

(132973, 19)

<IPython.core.display.Javascript object>

#### Extracting `age` from `info_0`

In [29]:
# For loop to extract age from info_0
for i, item in enumerate(df["info_0"]):
    pattern = r"\d"
    if re.search(pattern, item):
        df.loc[i, "age"] = int(item)

# Checking one of these rows
df[df["info_0"] == "92"]

day              name                                 info  \
33085   28    David Turnbull  . 92, American materials scientist.   
117386  11  Gotthilf Fischer        92, German choral conductor .   

                                                                      link  \
33085   https://en.wikipedia.org/wiki/David_Turnbull_(materials_scientist)   
117386                      https://en.wikipedia.org/wiki/Gotthilf_Fischer   

       num_references  year     month info_0                        info_1  \
33085               8  2007     April     92  American materials scientist   
117386              4  2020  December     92       German choral conductor   

       info_2 info_3 info_4 info_5 info_6 info_7 info_8 info_9 info_10  \
33085     NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN     NaN   
117386    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN     NaN   

       info_11   age  
33085      NaN  92.0  
117386     NaN  92.0

<IPython.core.display.Javascript object>

In [30]:
# Dropping info_0
df.drop("info_0", axis=1, inplace=True)

<IPython.core.display.Javascript object>

#### Observations:
- The new `age` column has been added successfully.
- We are finished processing `info_0`.

### `info_1`

#### Unique Values

In [31]:
# Checking unique values
df["info_1"].unique()

array(['86', '68', '87', '93', '79', '50', '88', '72', '81', '80', '90',
       '85', '92', '58', '54', '96', '49', '77', '76', '43', '35', '83',
       '31', '64', '57', '52', '84', '78', '70', '73', '67', '99', '33',
       '75', '66', '74', '62', '61', '82', '38', '47', '56', '91', '89',
       '94', '45', '65', '97', '63', '69', '37', '53', '46', '60', '26',
       '71', '25', '59', '39–40', '23', '95', '42', '32', '51', '41',
       '55', '44', '98', '39', '100', '27', '28', '40', '30', '48', '34',
       '29', '36', '111', '22', '104', '14', '21', '106', '105', '18',
       '101', '102', '20', '73-74', '84-85', '67-68', '85-86', '89-90',
       '103', '49-50', '87-88', '107', '48-49', '19', '71-72', '2',
       '82-83', '88-89', '11-12', '63-64', '64-65', '24', '17', '69-70',
       '52-53', '39-40', '80-81', '51-52', '70–71', '66–67', '59-60',
       '94-95', '15', '108', '76-77', '1995', '37-38',
       'American politician', '55/56', '87–88',
       '90 Swedish Olympic sprinte

<IPython.core.display.Javascript object>

#### Observations:
- There is a lot of variety in the format of the age data.
- Also, this field contains several values that we would expect in info_2 and beyond.
- Let us take the approach of extracting age values first, then examining rows with missing `age`.

#### Examining unique formats for age data

In [32]:
# Examining formats for age data
has_num = []
for i, item in enumerate(df["info_1"]):
    pattern = r"\d"
    if re.search(pattern, item) is not None:
        has_num.append(i)

df.loc[has_num, :]["info_1"].unique()

array(['86', '68', '87', '93', '79', '50', '88', '72', '81', '80', '90',
       '85', '92', '58', '54', '96', '49', '77', '76', '43', '35', '83',
       '31', '64', '57', '52', '84', '78', '70', '73', '67', '99', '33',
       '75', '66', '74', '62', '61', '82', '38', '47', '56', '91', '89',
       '94', '45', '65', '97', '63', '69', '37', '53', '46', '60', '26',
       '71', '25', '59', '39–40', '23', '95', '42', '32', '51', '41',
       '55', '44', '98', '39', '100', '27', '28', '40', '30', '48', '34',
       '29', '36', '111', '22', '104', '14', '21', '106', '105', '18',
       '101', '102', '20', '73-74', '84-85', '67-68', '85-86', '89-90',
       '103', '49-50', '87-88', '107', '48-49', '19', '71-72', '2',
       '82-83', '88-89', '11-12', '63-64', '64-65', '24', '17', '69-70',
       '52-53', '39-40', '80-81', '51-52', '70–71', '66–67', '59-60',
       '94-95', '15', '108', '76-77', '1995', '37-38', '55/56', '87–88',
       '90 Swedish Olympic sprinter', '6', '86-87', '62/63', '79

<IPython.core.display.Javascript object>

#### Observations:
- The data for age within `info_1` is in the following formats:  
    - single integer ("age", "age ", "age.")
    - range of 2 integers (separators '-', '–', '/', and ' or ')
    - range of 2 integers with only unit value for second number ('age1/age2-2nd-digit')
    - age in days or months ('age days', 'age-months')
    - estimates ('c. age', 'c.age',  'age?', 'ages' (e.g. 80s), "age+"
- There are some specific rows that need to be examined, with the following values for `info_1`:
    - 1995
    - 1996
    - 1997
    - German Olympic sailor [1]
    - Taiwanese failed assassin in the 3-19 shooting incident
    - 255
    - 176
    - the first wild bear in Germany in 170 years
    - c. 3500
    - common chimpanzee 55
    - Maltese 15
    - c.1000
    - Tree of the Year 150
- We will need to be strategic in the order in which we extract age from `info_1`.
- First, we will look at the atypical values listed above.

#### Examining Rows with Digits and Atypical Values for `info_2`

In [33]:
# List of atypical info_1 values for rows with digits
values_lst = [
    "1995",
    "1996",
    "1997",
    "German Olympic sailor [1]",
    "Taiwanese failed assassin in the 3-19 shooting incident",
    "255",
    "176",
    "the first wild bear in Germany in 170 years",
    "c. 3500",
    "common chimpanzee 55",
    "Maltese 15",
    "c.1000",
    "Tree of the Year 150",
]

df[df["info_1"].isin(values_lst)]

day                name  \
6046     9        Khaptad Baba   
6985     5  Wymberley D. Coerr   
8441    12        Moro Lorenzo   
10548   21       Pakoda Kadhar   
22722   17    Klaus Oldendorff   
25010   29      Chen Yi-hsiung   
29898   23             Adwaita   
30566   23             Harriet   
30591   26            Bear JJ1   
52996   16         The Senator   
55404    2              Oliver   
55455    5       Lucky Diamond   
86515    8  Pioneer Cabin Tree   
118228   5      Happy Man Tree   

                                                                                                                                       info  \
6046                                                                                                      , 1995, Nepalese spiritual saint.   
6985                                                                                              , 1995, American politician and diplomat.   
8441                                                                                      , 1996, Filipino basketball player and executive.   
10548                                                                                                                 , 1997, Indian actor.   
22722                                                                                                           , German Olympic sailor [1]   
25010                                                                            , Taiwanese failed assassin in the 3-19 shooting incident.   
29898                          , 255 , tortoise claimant for world's oldest animal, reputedly a former pet of General Clive, liver failure.   
30566   , 176, Galápagos tortoise believed to be the third oldest animal in the world and allegedly owned by Charles Darwin, heart failure.   
30591                                                                         , the first wild bear in Germany in 170 years, shot to death.   
52996                                                                    , c. 3500, American pond cypress tree, largest in the world, fire.   
55404                                           , common chimpanzee 55, Congolese-born  noted for his upright stature and humanlike traits.   
55455                                                                   , Maltese 15, American Guinness World Record dog,  holder , cancer.   
86515                                                                                  , c.1000, American giant sequoia tree, storm damage.   
118228                                                                             , Tree of the Year 150, British plane tree,  , cut down.   

                                                     link num_references  \
6046           https://en.wikipedia.org/wiki/Khaptad_Baba              6   
6985     https://en.wikipedia.org/wiki/Wymberley_D._Coerr              3   
8441           https://en.wikipedia.org/wiki/Moro_Lorenzo              5   
10548         https://en.wikipedia.org/wiki/Pakoda_Kadhar              1   
22722      https://en.wikipedia.org/wiki/Klaus_Oldendorff              1   
25010        https://en.wikipedia.org/wiki/Chen_Yi-hsiung              4   
29898               https://en.wikipedia.org/wiki/Adwaita              4   
30566    https://en.wikipedia.org/wiki/Harriet_(tortoise)             10   
30591              https://en.wikipedia.org/wiki/Bear_JJ1              1   
52996    https://en.wikipedia.org/wiki/The_Senator_(tree)             11   
55404   https://en.wikipedia.org/wiki/Oliver_(chimpanzee)             14   
55455   https://en.wikipedia.org/wiki/Lucky_Diamond_(dog)             13   
86515    https://en.wikipedia.org/wiki/Pioneer_Cabin_Tree             45   
118228       https://en.wikipedia.org/wiki/Happy_Man_Tree             23   

        year    month  \
6046    1996      May   
6985    1996  October   
8441    1997    April   
10548   1998  January   
22722   2003    March   
25010   2004    March   
29898   2006    March   
30566   2

<IPython.core.display.Javascript object>

#### Observations:
- Age data is either missing or the entry is for a member of a non-human species.
- We will drop all of these rows.

#### Dropping Rows for Non-Human Entries or Entries Missing Age Data

In [34]:
# List of indexes to be dropped
drop_rows = df[df["info_1"].isin(values_lst)].index.to_list()

# Dropping rows, resetting index, and checking new shape of df
df.drop(drop_rows, inplace=True)
df.reset_index(inplace=True, drop=True)
df.shape

(132959, 19)

<IPython.core.display.Javascript object>

#### Observations:
- With those rows addressed, we will begin extracting age in days and months and convert them to years.

### Extracting `age` from `info_1`

#### Step 1: Age in Years and Months

In [35]:
# Dictionary of patterns for days and months formats as keys and factor to convert to years
patterns = {
    r"(\d{1,3})( days)": 365,
    r"(\d{1,3})(-months)": 12,
    r"(\d{1,3})( months)": 12,
}

# For loop to check for age in days and months and convert to years and save in age.
# Also removes age in days/months from info_0.
for key, value in patterns.items():
    for i, item in enumerate(df["info_1"]):
        match = re.search(key, item)
        if match:
            age = int(match.group(1)) / value
            df.loc[i, "age"] = age
            df.loc[i, "info_1"] = re.sub(key, "", df.loc[i, "info_1"])

# Checking updated rows
df[df["age"].notna()]

day                         name  \
28491   12  Susan Anne Catherine Torres   
30523   18          Chris and Cru Kahui   
33076   28               David Turnbull   
59603   24          Kristján Jóhannsson   
90455   28                 Charlie Gard   
117373  11             Gotthilf Fischer   

                                                                                                            info  \
28491   , 40 days, American baby born to Susan Torres, brain-dead woman, heart failure after intestinal surgery.   
30523                                                            , 3-months, New Zealand child homicide victims.   
33076                                                                        . 92, American materials scientist.   
59603                                                                             83, Icelandic Olympic athlete.   
90455                       , 11 months, British infant, subject of life support and parental rights case, MDDS.   
117373                                                                             92, German choral conductor .   

                                                                         link  \
28491                              https://en.wikipedia.org/wiki/Susan_Torres   
30523                       https://en.wikipedia.org/wiki/Chris_and_Cru_Kahui   
33076      https://en.wikipedia.org/wiki/David_Turnbull_(materials_scientist)   
59603   https://en.wikipedia.org/wiki/Kristj%C3%A1n_J%C3%B3hannsson_(athlete)   
90455                         https://en.wikipedia.org/wiki/Charlie_Gard_case   
117373                         https://en.wikipedia.org/wiki/Gotthilf_Fischer   

       num_references  year      month                        info_1  \
28491               3  2005  September                                 
30523              48  2006       June                                 
33076               8  2007      April  American materials scientist   
59603               2  2013    January     Icelandic Olympic athlete   
90455             125  2017       July                                 
117373              4  2020   December       German choral conductor   

                                    info_2  \
28491   American baby born to Susan Torres   
30523   New Zealand child homicide victims   
33076                                  NaN   
59603                                  NaN   
90455                       British infant   
117373                                 NaN   

                                                  info_3  \
28491                                   brain-dead woman   
30523                                                NaN   
33076                                                NaN   
59603                                                NaN   
90455   subject of life support and parental rights case   
117373                                               NaN   

                                        info_4 info_5 info_6 info_7 info_8  \
28491   heart failure after intestinal surgery    NaN    NaN    NaN    NaN   
30523                                      NaN    NaN    NaN    NaN    NaN   
33076                                      NaN    NaN    NaN    NaN    NaN   
59603                                      NaN    NaN    NaN    NaN    NaN   
90455                                     MDDS    NaN    NaN    NaN    NaN   
117373                                     NaN    NaN    NaN    NaN    NaN   

       info_9 info_10 info_11        age  
28491     NaN     NaN     NaN   0.109589  
30523     NaN     NaN     NaN   0.250000  
33076     NaN     NaN     NaN  92.000000  
59603     NaN     NaN     NaN  83.000000  
90455     NaN     NaN     NaN   0.916667  
117373    NaN     NaN     NaN  92.000000

<IPython.core.display.Javascript object>

#### Observations:
- We have successfully captured the age in days and months values and converted them to years.
- The other rows are in place already from our treatment of `info_0`.
- Next, we will address entries that contain two age values as a range, starting first with those that have a single digit as the second number.

#### Step 2: Extracting `age` from `info_1` for Entries with Age Estimate Containing Age Range with 2 Values

#### Ranges with Single Digit as Upper-end

In [36]:
# Pattern for re
pattern = r"(\d{1,3})(/)(\d)\b"

# List of rows to check
rows_to_check = []

# For loop to add index of list containing values to list
for i, item in enumerate(df["info_1"]):
    match = re.search(pattern, item)
    if match:
        rows_to_check.append(i)

# Checking sample of rows
df.loc[rows_to_check, :].sample(2)

day           name  \
27405  12  Nelly Uchendu   
40905  17    Joji Banuve   

                                                                                                     info  \
27405                                                                          , 54/5, Nigerian musician.   
40905  , 68/9, Fijian politician, Minister for Local Government and the Environment, after short illness.   

                                              link num_references  year  \
27405  https://en.wikipedia.org/wiki/Nelly_Uchendu             11  2005   
40905    https://en.wikipedia.org/wiki/Joji_Banuve              0  2009   

       month info_1             info_2  \
27405  April   54/5  Nigerian musician   
40905   June   68/9  Fijian politician   

                                                  info_3               info_4  \
27405                                                NaN                  NaN   
40905  Minister for Local Government and the Environment  after short illness   

      info_5 info_6 info_7 info_8 info_9 info_10 info_11  age  
27405    NaN    NaN    NaN    NaN    NaN     NaN     NaN  NaN  
40905    NaN    NaN    NaN    NaN    NaN     NaN     NaN  NaN

<IPython.core.display.Javascript object>

In [37]:
# Pattern for re
pattern = r"(\d{1,3})(/)(\d)\b"


# For loop to find rows with values and pattern and extract and calculate age to age column and remove age from info_1
for i in df[df["age"].isna()].index:
    item = df.loc[i, "info_1"]
    match = re.search(pattern, item)
    if match:
        age_1 = int(match.group(1))
        age_2 = int(match.group(3))
        units = ((age_1 % 10) + age_2) / 2
        tens = age_1 - (age_1 % 10)
        age = tens + units
        df.loc[i, "age"] = age
        df.loc[i, "info_1"] = re.sub(pattern, "", df.loc[i, "info_1"])

# Checking example rows
pd.concat([df[df["name"] == "Lu'ay al-Atassi"], df[df["name"] == "Joji Banuve"]])

day             name  \
24271  24  Lu'ay al-Atassi   
40905  17      Joji Banuve   

                                                                                                     info  \
24271                                           , 76/7, Syrian army commander and politician, President .   
40905  , 68/9, Fijian politician, Minister for Local Government and the Environment, after short illness.   

                                                  link num_references  year  \
24271  https://en.wikipedia.org/wiki/Lu%27ay_al-Atassi              6  2003   
40905        https://en.wikipedia.org/wiki/Joji_Banuve              0  2009   

          month info_1                                info_2  \
24271  November         Syrian army commander and politician   
40905      June                            Fijian politician   

                                                  info_3               info_4  \
24271                                          President                  NaN   
40905  Minister for Local Government and the Environment  after short illness   

      info_5 info_6 info_7 info_8 info_9 info_10 info_11   age  
24271    NaN    NaN    NaN    NaN    NaN     NaN     NaN  76.5  
40905    NaN    NaN    NaN    NaN    NaN     NaN     NaN  68.5

<IPython.core.display.Javascript object>

#### Other Ranges with Two Values

In [38]:
# Pattern for re
pattern = r"(\d{1,3})(-|–|/| or )(\d{1,3})"

# List of rows to check
rows_to_check = []

# For loop to add index of list containing values to list
for i, item in enumerate(df["info_1"]):
    match = re.search(pattern, item)
    if match:
        rows_to_check.append(i)

# Checking sample of rows
df.loc[rows_to_check, :].sample(2)

day                 name  \
130389  16  Esmaeil Jabbarzadeh   
115201  13       Said Ali Kemal   

                                                                                      info  \
130389                                   , 61–62, Iranian politician, MP , cardiac arrest.   
115201  , 81–82, Comorian politician, Member of the Assembly of the Union of the Comoros .   

                                                     link num_references  \
130389  https://en.wikipedia.org/wiki/Esmaeil_Jabbarzadeh              1   
115201       https://en.wikipedia.org/wiki/Said_Ali_Kemal              6   

        year      month info_1               info_2  \
130389  2022   February  61–62   Iranian politician   
115201  2020  September  81–82  Comorian politician   

                                                    info_3          info_4  \
130389                                                  MP  cardiac arrest   
115201  Member of the Assembly of the Union of the Comoros             NaN   

       info_5 info_6 info_7 info_8 info_9 info_10 info_11  age  
130389    NaN    NaN    NaN    NaN    NaN     NaN     NaN  NaN  
115201    NaN    NaN    NaN    NaN    NaN     NaN     NaN  NaN

<IPython.core.display.Javascript object>

In [39]:
# Pattern for re
pattern = r"(\d{1,3})(-|–|/| or )(\d{1,3})"


# For loop to find rows with values and pattern and extract age to age column and remove age from info_1
for i in df[df["age"].isna()].index:
    item = df.loc[i, "info_1"]
    match = re.search(pattern, item)
    if match:
        age = (int(match.group(1)) + int(match.group(3))) / 2
        df.loc[i, "age"] = age
        df.loc[i, "info_1"] = re.sub(pattern, "", df.loc[i, "info_1"])

# Checking example rows
pd.concat([df[df["name"] == "Moncef Ouannes"], df[df["name"] == "Fehmi Sağınoğlu"]])

day             name                                      info  \
116353   4   Moncef Ouannes  , 63–64, Tunisian sociologist, COVID-19.   
84500   19  Fehmi Sağınoğlu              , 78/79, Turkish footballer.   

                                                                link  \
116353                  https://en.wikipedia.org/wiki/Moncef_Ouannes   
84500   https://en.wikipedia.org/wiki/Fehmi_Sa%C4%9F%C4%B1no%C4%9Flu   

       num_references  year      month info_1                info_2    info_3  \
116353              7  2020   November         Tunisian sociologist  COVID-19   
84500               2  2016  September           Turkish footballer       NaN   

       info_4 info_5 info_6 info_7 info_8 info_9 info_10 info_11   age  
116353    NaN    NaN    NaN    NaN    NaN    NaN     NaN     NaN  63.5  
84500     NaN    NaN    NaN    NaN    NaN    NaN     NaN     NaN  78.5

<IPython.core.display.Javascript object>

#### Observations:
- Next, we will extract the entries with straightforward single integer age values, including the formats: "age", "age ", 'age.'.
- More vague estimates are excluded here to allow closer examination, as they are more likely to by atypical entries.

#### Step 3: Age as Single Integer (Excluding Estimates)

In [40]:
# List of patterns for age formats with single integer for age
patterns = [r"^(\d{1,3})$", r"^(\d{1,3})\s", r"^(\d{1,3})\.\s"]

# For loop to check age pattern in info_0, save age to age column, and remove from age from info_0
for i, item in enumerate(df["info_1"]):
    for pattern in patterns:
        match = re.search(pattern, item)
        if match:
            age = int(match.group(1))
            df.loc[i, "age"] = age
            df.loc[i, "info_1"] = re.sub(pattern, "", df.loc[i, "info_1"])

# Checking first 2 rows
df.head(2)

day              name                                                 info  \
0   1  William Chappell  , 86, British dancer, ballet designer and director.   
1   1    Raymond Crotty         , 68, Irish economist, writer, and academic.   

                                                      link num_references  \
0  https://en.wikipedia.org/wiki/William_Chappell_(dancer)             21   
1             https://en.wikipedia.org/wiki/Raymond_Crotty             12   

   year    month info_1           info_2                        info_3  \
0  1994  January          British dancer  ballet designer and director   
1  1994  January         Irish economist                        writer   

         info_4 info_5 info_6 info_7 info_8 info_9 info_10 info_11   age  
0           NaN    NaN    NaN    NaN    NaN    NaN     NaN     NaN  86.0  
1  and academic    NaN    NaN    NaN    NaN    NaN     NaN     NaN  68.0

<IPython.core.display.Javascript object>

In [41]:
# Checking last 2 rows
df.tail(2)

day           name  \
132957   9  Oleg Moliboga   
132958   9       Zou Jing   

                                                                         info  \
132957          , 69, Russian volleyball player, Olympic champion  and coach.   
132958  , 86, Chinese engineer, member of the Chinese Academy of Engineering.   

                                                     link num_references  \
132957        https://en.wikipedia.org/wiki/Oleg_Moliboga              2   
132958  https://en.wikipedia.org/wiki/Zou_Jing_(engineer)              3   

        year month info_1                     info_2  \
132957  2022  June         Russian volleyball player   
132958  2022  June                  Chinese engineer   

                                              info_3 info_4 info_5 info_6  \
132957                   Olympic champion  and coach    NaN    NaN    NaN   
132958  member of the Chinese Academy of Engineering    NaN    NaN    NaN   

       info_7 info_8 info_9 info_10 info_11   age  
132957    NaN    NaN    NaN     NaN     NaN  69.0  
132958    NaN    NaN    NaN     NaN     NaN  86.0

<IPython.core.display.Javascript object>

In [42]:
# Checking the number of remaining missing values for `age`
df["age"].isna().sum()

294

<IPython.core.display.Javascript object>

#### Observations:
- The rows with single integer age data have been addressed.
- There are 294 remaining missing values for `age`.
- Let us check the rows containing 'c.', '+', or '?' in the age information.

#### Entries with Age Data Containing 'c.', '+', or '?' for Estimate

In [43]:
values = ["c.", "+", "?"]

# List of rows to check
rows_to_check = []

# For loop to add index of list containing values to list
for index in df[df["age"].isna()].index:
    if any(value in df.loc[index, "info_1"] for value in values):
        rows_to_check.append(index)

# Inspecting rows containing values
df.loc[rows_to_check, :]

day                                  name  \
12921   22                         Graciela Quan   
25365    4                   Tesfaye Gebre Kidan   
25445   18                          Paul Johnson   
25446   18                          Nek Mohammed   
26726   13                          Earl Cameron   
29743   28                            Peter Snow   
29839   15                              Humphrey   
30258   14                            Paul Marco   
31173    6          Mohammed Taha Mohammed Ahmed   
31455   11                       Benito Martínez   
32767   21                       Mohd. Ayub Khan   
33116    5     Prince Abdul-Majid bin Abdul-Aziz   
33989   21                             Čabulītis   
34571   30                                Washoe   
34625    6                  Sayed Mustafa Kazemi   
38029    1                        Arlene Sherman   
39779    2                         Halit Balamir   
40777    3                           Moloko Temo   
41419    4                                Benson   
41438    5  Sheikha Hessa bint Salman Al Khalifa   
41439    5                      Baitullah Mehsud   
41895   14                Saleh Ali Saleh Nabhan   
43569   28              Mohammad-Reza Ali-Zamani   
43587   28                     Arash Rahmanipour   
44021   27                     Black Bear Island   
44098    4                             Big Truck   
44113    5                        Aminah Assilmi   
45202   24                       Abdolhamid Rigi   
47375   15                        Nimr al-Khatib   
48566   18                             Bob Tanna   
49230   12                         Buster Martin   
49490    1                            Anny Rüegg   
49983    8               Fazul Abdullah Mohammed   
53782   28                Abukar Hassan Mohamoud   
54951    5                                  Meow   
55770   24                       Lonesome George   
56359   31           Abdi Jeylani Malaq Marshale   
58605    9                              Ataa Oko   
61244   17                             Bi Kidude   
61386   23                         Mohammed Omar   
62112    4                           Luo Meizhen   
62546   27                      Muhammad Emin Er   
64750    3                                Reshma   
66409   30                               Greater   
67511   22                    Hamza Abu al-Haija   
69857   22                          Glenn Jowitt   
70432   21                          Raed al Atar   
71115   26                                 Zelda   
73700    4      Sajida Mubarak Atrous al-Rishawi   
74506   19             Leandra Becerra Lumbreras   
74710   28                         Tuti Yusupova   
75566   13                           Bob Randall   
75796   27        Hassan Abdullah Hersi al-Turki   
78508    2                        Omar El-Hariri   
82377   21                        Akhtar Mansour   
92358   14                           Little Mama   
95095   21                            Kakon Bibi   
100115  17                        Zura Karuhimbi   
105358  19                        Carlos Porrata   
109857  13                        Nasser Shabani   
128544  20                     Mamadú Iaia Djaló   

                                                                                                                           info  \
12921                                                                   , c. 79, Guatemalan lawyer and women's rights activist.   
25365                                                   , c. 69, Ethiopian general, defense minister and President of Ethiopia.   
25445                                                                       , c. 49, American hostage, decapitated by al-Qaeda.   
25446                 , c. 27, Pakistani tribal leader in Waziristan and key Taliban ally, killed by Pakistani military forces.   
26726                                                                                 , 89?, Canadian broadcaster and  anchor .

<IPython.core.display.Javascript object>

#### Observations:
- Most of the entries are for people, but there are also one or more entries for the following:
    - carp
    - racehorse
    - chimpanzee
    - flamingo
    - cat
    - turkey
- We will proceed to check `info_2` for these values and drop these and other rows representing members of these other species.

#### Checking for Cat, Racehorse, Chimpanzee, Carp, and Flamingo in `info_2`

In [44]:
# Defining pattern for re
species_pattern = r"\b(cat|racehorse|chimpanzee|carp|flamingo|turkey)\b"

# Empty list to collect indexes of rows containing pattern
rows_to_check = []

# For loop to add index of rows containing pattern to list
for index in df[df["info_2"].notna()].index:
    match = re.search(species_pattern, df.loc[index, "info_2"])
    if match:
        rows_to_check.append(index)

# Checking number of rows containing pattern
print(f"There are {len(rows_to_check)} rows containing these values.")

There are 468 rows containing these values.


<IPython.core.display.Javascript object>

#### Observations:
- There are sufficient rows to warrant checking species by species.

#### Cat Entries per `info_2`

In [45]:
# Defining pattern for re
species_pattern = r"\b(cat)\b"

# Empty list to collect indexes of rows containing pattern
rows_to_check = []

# For loop to add index of rows containing pattern to list
for index in df[df["info_2"].notna()].index:
    match = re.search(species_pattern, df.loc[index, "info_2"])
    if match:
        rows_to_check.append(index)

# Inpsecting rows with pattern
df.loc[rows_to_check, :]

day                  name  \
18610    4         Lorenzo Music   
31855   29  Dewey Readmore Books   
38127   11              Scarlett   
39070    4                 India   
39688   20                 Socks   
41343   27                 Sybil   
47436   21          Prince Chunk   
54951    5                  Meow   
59824    4                Stewie   
63271    7     Buurtpoes Bledder   
66389   29          Colonel Meow   
70837   11       Hamish McHamish   
72423    4       Frank and Louie   
76283   22                  Tama   
90321   21                Stubbs   
92237    7               Paddles   
103398  14            Grumpy Cat   
107382   1               Lil Bub   
109613   3                    CC   

                                                                                                                              info  \
18610   , 64, American voice actor known for the voice of the cartoon cat Garfield, complications related to lung and bone cancer.   
31855                                                                                               , 19,  Library cat, euthanized   
38127                           , 13, American stray cat, name source of the Scarlett Award for Animal Heroism, animal euthanasia.   
39070                                                                                    , 18, American pet cat of George W. Bush.   
39688                                                           , 19, American Presidential cat of the Clinton family, euthanized.   
41343                                   , 3, British Downing Street cat, Chief Mouser to the Cabinet Office , after short illness.   
47436                                                                                     , 10, American obese cat, heart disease.   
54951                                                       , c. 2, American cat, heaviest cat at his time of death, lung failure.   
59824                                                                                 , 7–8, world's longest domestic cat, cancer.   
63271                                                                          , 1–2, domestic cat in Netherlands, motor accident.   
66389                                           , 2, American Himalayan-Persian cat, Guinness World Record holder for longest fur.   
70837                                                                               , 15, Scottish celebrity cat, chest infection.   
72423                                                  , 15, American-born Ragdoll cat, world's oldest diprosopus cat, euthanized.   
76283                                                    , 16, Japanese calico cat, stationmaster at Kishi Station, heart failure.   
90321                                                                    , 20, American cat, honorary mayor of Talkeetna, Alaska .   
92237                               , 1, New Zealand polydactyl cat, co-owned by Prime Minister Jacinda Ardern, traffic collision.   
103398                                           , 7, American internet celebrity cat, complications from urinary tract infection.   
107382                                                                                               , 8, American celebrity cat .   
109613                                                                 , 18, American tabby cat, first cloned pet, kidney failure.   

                                                      link num_references  \
18610          https://en.wikipedia.org/wiki/Lorenzo_Music              6   
31855   https://en.wikipedia.org/wiki/Dewey_Readmore_Books             26   
38127         https://en.wikipedia.org/wiki/Scarlett_(cat)              6   
39070            https://en.wikipedia.org/wiki/India_(cat)              7   
39688            https://en.wikipedia.org/wiki/Socks_(cat)             21   
41343            https://en.wikipedia.org/wiki/Sybil_(cat)              5   
47436           https://en.wikipedia.org/wiki/Prince_Chunk        

<IPython.core.display.Javascript object>

#### Observations:
- There is one person represented.  
- We can proceed to drop the others, using "cat", "cat of", and "cat in" patterns.

#### Dropping Entries for Cats per `info_2`

In [46]:
# List of re patterns to find
patterns = [r"\bcat$", r"\b(cat of|cat in)\b"]

# List to collect indexes of rows to drop
rows_to_drop = []

# For loop to find re pattern and add index of rows with pattern to list
for index in df[df["info_2"].notna()].index:
    for pattern in patterns:
        match = re.search(pattern, df.loc[index, "info_2"])
        if match:
            rows_to_drop.append(index)

# Checking number of rows added to list for dropping
print(f"{len(rows_to_drop)} rows will be dropped.")

18 rows will be dropped.


<IPython.core.display.Javascript object>

In [47]:
# Dropping rows, resetting index, and checking new shape of df
df.drop(rows_to_drop, inplace=True)
df.reset_index(inplace=True, drop=True)
df.shape

(132941, 19)

<IPython.core.display.Javascript object>

####  Racehorse Entries per `info_2`

In [48]:
# Defining pattern for re
pattern = r"\bracehorse\b"

# Empty list to collect indexes of rows containing pattern
rows_to_check = []

# For loop to add index of rows containing pattern to list
for index in df[df["info_2"].notna()].index:
    match = re.search(pattern, df.loc[index, "info_2"])
    if match:
        rows_to_check.append(index)

# Inpsecting rows with pattern
df.loc[rows_to_check, :].sample(10)

day                 name  \
35750   17       Winning Colors   
65337    8                Jwala   
39983   21          Beach Towel   
17471   25  John J. Tammaro Jr.   
87299   13         Lucky Pulpit   
109329  21            A.P. Indy   
99628   20       Monet's Garden   
102660   8              Turgeon   
99697   24       I Want Revenge   
71288    7                Cigar   

                                                                                            info  \
35750                             , 23, American racehorse, won 1988 Kentucky Derby, euthanized.   
65337                                           , 4, British Thoroughbred racehorse, euthanised.   
39983                       , 22, American harness racehorse, Harness Horse of the Year , colic.   
17471                                             , 75, American Thoroughbred racehorse trainer.   
87299                                                    , 16, American racehorse, heart attack.   
109329                                                        , 31, American racehorse and sire.   
99628                                                         , 20, Irish racehorse, euthanized.   
102660                                         , 33, French racehorse, Cartier Champion Stayer .   
99697                                              , 12, American Thoroughbred racehorse, virus.   
71288   , 24, American thoroughbred racehorse, complications of neck surgery for osteoarthritis.   

                                                          link num_references  \
35750     https://en.wikipedia.org/wiki/Winning_Colors_(horse)             11   
65337              https://en.wikipedia.org/wiki/Jwala_(horse)             15   
39983        https://en.wikipedia.org/wiki/Beach_Towel_(horse)              6   
17471        https://en.wikipedia.org/wiki/John_J._Tammaro_Jr.              0   
87299               https://en.wikipedia.org/wiki/Lucky_Pulpit             14   
109329                 https://en.wikipedia.org/wiki/A.P._Indy             31   
99628   https://en.wikipedia.org/wiki/Monet%27s_Garden_(horse)              6   
102660           https://en.wikipedia.org/wiki/Turgeon_(horse)              1   
99697             https://en.wikipedia.org/wiki/I_Want_Revenge              7   
71288              https://en.wikipedia.org/wiki/Cigar_(horse)             51   

        year     month info_1                                   info_2  \
35750   2008  February                              American racehorse   
65337   2013  December                  British Thoroughbred racehorse   
39983   2009     March                      American harness racehorse   
17471   2001  February         American Thoroughbred racehorse trainer   
87299   2017  February                              American racehorse   
109329  2020  February                     American racehorse and sire   
99628   2018  November                                 Irish racehorse   
102660  2019     April                                French racehorse   
99697   2018  November                 American Thoroughbred racehorse   
71288   2014   October                 American thoroughbred racehorse   

                                                  info_3      info_4 info_5  \
35750                            won 1988 Kentucky Derby  euthanized    NaN   
65337                                         euthanised         NaN    NaN   
39983                          Harness Horse of the Year       colic    NaN   
17471                                                NaN         NaN    NaN   
87299                                       heart attack         NaN    NaN   
109329                                               NaN         NaN    NaN   
99628                                         euthanized         NaN    NaN   
102660                           Cartier Champion Stayer         NaN    NaN   
99697                                              virus         NaN    NaN   
71288   complications o

<IPython.core.display.Javascript object>

#### Observations:
- There are several entries for people involved in the racehorse business.
- Values that end in 'racehorse' and 'racehorse and sire' can be removed.

#### Dropping Entries for Racehorses per `info_2`

In [49]:
# List of re patterns to find
patterns = [r"\bracehorse$", r"\b(racehorse and sire)$"]

# List to collect indexes of rows to drop
rows_to_drop = []

# For loop to find re pattern and add index of rows with pattern to list
for index in df[df["info_2"].notna()].index:
    for pattern in patterns:
        match = re.search(pattern, df.loc[index, "info_2"])
        if match:
            rows_to_drop.append(index)

# Checking number of rows added to list for dropping
print(f"{len(rows_to_drop)} rows will be dropped.")

349 rows will be dropped.


<IPython.core.display.Javascript object>

In [50]:
# Dropping rows, resetting index, and checking new shape of df
df.drop(rows_to_drop, inplace=True)
df.reset_index(inplace=True, drop=True)
df.shape

(132592, 19)

<IPython.core.display.Javascript object>

####  Chimpanzee, Flamingo,  Carp, and Turkey Entries per `info_2`

In [51]:
# Defining pattern for re
pattern = r"\b(chimpanzee|flamingo|carp|turkey)\b"

# Empty list to collect indexes of rows containing pattern
rows_to_check = []

# For loop to add index of rows containing pattern to list
for index in df[df["info_2"].notna()].index:
    match = re.search(pattern, df.loc[index, "info_2"])
    if match:
        rows_to_check.append(index)

# Inpsecting rows with pattern
df.loc[rows_to_check, :]

day                 name  \
34556  30               Washoe   
39598  16               Travis   
41364   4               Benson   
45230   1  Heather the Leather   
66268  30              Greater   
70952  26                Zelda   
76075  22     Don Featherstone   
92090  14          Little Mama   

                                                                                                                        info  \
34556  , c.42, American-trained African-born chimpanzee believed to be first non-human to acquire human language, influenza.   
39598                                                    , 13, American-born chimpanzee, television commercial animal, shot.   
41364                                                      , c. 25, British common carp, voted as Britain's Favourite Carp .   
45230                                                                                 , 50, British scaleless carp, old age.   
66268                                             , c. 83, Australian greater flamingo, world's oldest flamingo, euthanized.   
70952                              , 11+, American wild turkey, resident of New York City's Battery Park, traffic collision.   
76075                                   , 79, American artist and inventor of the plastic pink flamingo, Lewy body dementia.   
92090                                                    , c.79, African-born chimpanzee , oldest on record, kidney failure.   

                                                          link num_references  \
34556        https://en.wikipedia.org/wiki/Washoe_(chimpanzee)             42   
39598        https://en.wikipedia.org/wiki/Travis_(chimpanzee)             69   
41364              https://en.wikipedia.org/wiki/Benson_(fish)              9   
45230        https://en.wikipedia.org/wiki/Heather_the_Leather              5   
66268         https://en.wikipedia.org/wiki/Greater_(flamingo)             12   
70952             https://en.wikipedia.org/wiki/Zelda_(turkey)             13   
76075  https://en.wikipedia.org/wiki/Don_Featherstone_(artist)             10   
92090                https://en.wikipedia.org/wiki/Little_Mama              3   

       year      month info_1  \
34556  2007    October   c.42   
39598  2009   February          
41364  2009     August  c. 25   
45230  2010       June          
66268  2014    January  c. 83   
70952  2014  September    11+   
76075  2015       June          
92090  2017   November   c.79   

                                                                                                  info_2  \
34556  American-trained African-born chimpanzee believed to be first non-human to acquire human language   
39598                                                                           American-born chimpanzee   
41364                                                                                British common carp   
45230                                                                             British scaleless carp   
66268                                                                        Australian greater flamingo   
70952                                                                               American wild turkey   
76075                                          American artist and inventor of the plastic pink flamingo   
92090                                                                            African-born chimpanzee   

                                         info_3             info_4 info_5  \
34556                                 influenza                NaN    NaN   
39598              television commercial animal               shot    NaN   
41364         voted as Britain's Favourite Carp                NaN    NaN   
45230                                   old age                NaN    NaN   
66268                   world's oldest flamingo         euthanized    NaN   
70952  resident of New York City's Battery Park  traffic collision    NaN   
76075 

<IPython.core.display.Javascript object>

#### Observations:
- All of these rows are for animals, so we will remove them.

#### Dropping Entries for Chimpanzees, Flamingos,  Carps, and Turkeys per `info_2`

In [52]:
# List of re patterns to find
patterns = [r"\b(chimpanzee|flamingo|carp|turkey)\b"]

# List to collect indexes of rows to drop
rows_to_drop = []

# For loop to find re pattern and add index of rows with pattern to list
for index in df[df["info_2"].notna()].index:
    for pattern in patterns:
        match = re.search(pattern, df.loc[index, "info_2"])
        if match:
            rows_to_drop.append(index)

# Checking number of rows added to list for dropping
print(f"{len(rows_to_drop)} rows will be dropped.")

8 rows will be dropped.


<IPython.core.display.Javascript object>

In [53]:
# Dropping rows, resetting index, and checking new shape of df
df.drop(rows_to_drop, inplace=True)
df.reset_index(inplace=True, drop=True)
df.shape

(132584, 19)

<IPython.core.display.Javascript object>

#### Observations:
- With those non-human entries addressed, we can return to processing `info_`.
- Let us address the remaining entries with '?', '+', or 'c.', accepting the estimated age as `age`.
- We will treat the age estimates ending in 's' similarly, but separately.

#### Extracting `age` from `info_1` for Entries with Age Estimate Containing '?', '+', or 'c.'

In [54]:
# List to identify rows
values = ["c.", "+", "?"]

# Pattern for re
pattern = r"\b(\d{1,3})\b"

# For loop to find rows with values and pattern and extract age to age column and remove age from info_1
for i in df[df["age"].isna()].index:
    item = df.loc[i, "info_1"]

    if any(value in item for value in values):
        match = re.search(pattern, item)

        if match:
            age = int(match.group(1))
            df.loc[i, "age"] = age
            df.loc[i, "info_1"] = re.sub(pattern, "", df.loc[i, "info_1"])

        for value in values:
            df.loc[i, "info_1"] = df.loc[i, "info_1"].replace(value, "")

# Checking example rows
pd.concat([df[df["name"] == "Tuti Yusupova"], df[df["name"] == "Raed al Atar"]])

day           name  \
74516  28  Tuti Yusupova   
70267  21   Raed al Atar   

                                                                            info  \
74516  , 134?, Uzbekistani longevity claimant, unverified world's oldest person.   
70267                            , c. 40, Palestinian militant , missile strike.   

                                              link num_references  year  \
74516  https://en.wikipedia.org/wiki/Tuti_Yusupova              5  2015   
70267   https://en.wikipedia.org/wiki/Raed_al_Atar             19  2014   

        month info_1                          info_2  \
74516   March         Uzbekistani longevity claimant   
70267  August                   Palestinian militant   

                                 info_3 info_4 info_5 info_6 info_7 info_8  \
74516  unverified world's oldest person    NaN    NaN    NaN    NaN    NaN   
70267                    missile strike    NaN    NaN    NaN    NaN    NaN   

      info_9 info_10 info_11    age  
74516    NaN     NaN     NaN  134.0  
70267    NaN     NaN     NaN   40.0

<IPython.core.display.Javascript object>

#### Observations:
- The ages ending in 's' should be the only ones remaining in the `info_1` column.
- We will treat those now, putting them at the mid-range for the decade indicated.

#### Extracting `age` from `info_1` for Entries with Age Estimate ending in 's'

In [55]:
# Defining pattern for re
pattern = r"\b(\d{1,3})s\b"

# Empty list to collect indexes of rows containing pattern
rows_to_check = []

# For loop to add index of rows containing pattern to list
for index in df[df["age"].isna()].index:
    match = re.search(pattern, df.loc[index, "info_1"])
    if match:
        rows_to_check.append(index)

# Inpsecting rows with pattern
df.loc[rows_to_check, :]

day                   name  \
27469   22              Mary Dann   
28056   14          Jacques Roche   
33053   27           Magda Gerber   
41695    2      Abdullah Laghmani   
47970    9     Makinti Napanangka   
61930    1     Dorothy Napangardi   
62092   10          Timothy Apiyo   
71567   30  Mohamed Sheikh Ismail   
84772   20          Achieng Abura   
85973   27  Abu Jandal al-Kuwaiti   
87185   19       Shibaji Banerjee   
100951   4            Bunny Brown   
126281   9               Ali Atwa   
128087  17    Dah Sagbadjou Glele   
130871  19              Tom Moody   

                                                                                   info  \
27469                                            , early 80s, American Indian activist.   
28056                                                  , early 40s, Haitian journalist.   
33053                                          , 90s, Hungarian-born American educator.   
41695                                   , 40s, Afghan Secret Service chief, bomb blast.   
47970                                            , 80s, Australian Papunya Tula artist.   
61930                           , 60s, Australian indigenous artist, traffic collision.   
62092                                    , 70s, Tanzanian politician and civil servant.   
71567                     , 50s, Somali military commander, Chief of the Police Force .   
84772                                                           , 50s, Kenyan musician.   
85973                                                    , 30s, Kuwaiti ISIL commander.   
87185                                          , 60s, Indian footballer , heart attack.   
100951                                    , 60s, Jamaican reggae musician, bone cancer.   
126281                                               , 60s, Lebanese hijacker , cancer.   
128087                                         , 90s, Beninese royal, king of Dahomey .   
130871  , 60s, American visual artist, critic and blogger, complications from COVID-19.   

                                                           link  \
27469   https://en.wikipedia.org/wiki/Mary_Dann_and_Carrie_Dann   
28056               https://en.wikipedia.org/wiki/Jacques_Roche   
33053                https://en.wikipedia.org/wiki/Magda_Gerber   
41695           https://en.wikipedia.org/wiki/Abdullah_Laghmani   
47970          https://en.wikipedia.org/wiki/Makinti_Napanangka   
61930          https://en.wikipedia.org/wiki/Dorothy_Napangardi   
62092               https://en.wikipedia.org/wiki/Timothy_Apiyo   
71567       https://en.wikipedia.org/wiki/Mohamed_Sheikh_Ismail   
84772               https://en.wikipedia.org/wiki/Achieng_Abura   
85973       https://en.wikipedia.org/wiki/Abu_Jandal_al-Kuwaiti   
87185            https://en.wikipedia.org/wiki/Shibaji_Banerjee   
100951                https://en.wikipedia.org/wiki/Bunny_Brown   
126281                   https://en.wikipedia.org/wiki/Ali_Atwa   
128087        https://en.wikipedia.org/wiki/Dah_Sagbadjou_Glele   
130871         https://en.wikipedia.org/wiki/Tom_Moody_(artist)   

       num_references  year      month     info_1  \
27469              12  2005      April  early 80s   
28056              25  2005       July  early 40s   
33053               7  2007      April        90s   
41695               4  2009  September        40s   
47970              57  2011    January        80s   
61930              19  2013       June        60s   
62092               2  2013       June        70s   
71567               8  2014    October        50s   
84772              13  2016    October        50s   
85973              16  2016   December        30s   
87185               1  2017   February        60s   
100951              2  2019   February        60s   
126281              3  2021    October        60s   
128087              8  2021   December        90s   
130871             10  2022      March        60s   

                                   

<IPython.core.display.Javascript object>

In [57]:
# Pattern for re
pattern = r"\b(\d{1,3})s\b"

# For loop to find rows with values and pattern and extract age to age column and remove age from info_1
for i in df[df["age"].isna()].index:
    item = df.loc[i, "info_1"]
    match = re.search(pattern, item)
    if match:
        age = int(match.group(1))
        df.loc[i, "age"] = age + 5
        df.loc[i, "info_1"] = re.sub(pattern, "", df.loc[i, "info_1"])
    if "early " in item:
        df.loc[i, "info_1"] = df.loc[i, "info_1"].replace("early ", "")

# Checking example rows
pd.concat([df[df["name"] == "Mary Dann"], df[df["name"] == "Timothy Apiyo"]])

day           name                                            info  \
27469  22      Mary Dann          , early 80s, American Indian activist.   
62092  10  Timothy Apiyo  , 70s, Tanzanian politician and civil servant.   

                                                          link num_references  \
27469  https://en.wikipedia.org/wiki/Mary_Dann_and_Carrie_Dann             12   
62092              https://en.wikipedia.org/wiki/Timothy_Apiyo              2   

       year  month info_1                                  info_2 info_3  \
27469  2005  April                       American Indian activist    NaN   
62092  2013   June         Tanzanian politician and civil servant    NaN   

      info_4 info_5 info_6 info_7 info_8 info_9 info_10 info_11   age  
27469    NaN    NaN    NaN    NaN    NaN    NaN     NaN     NaN  85.0  
62092    NaN    NaN    NaN    NaN    NaN    NaN     NaN     NaN  75.0

<IPython.core.display.Javascript object>

### Checking for Any Missed Digits in `info_1` and for Remaining Missing Values for `age`

In [70]:
pattern = r"\d"

rows_to_check = []
for i, item in enumerate(df["info_1"]):
    match = re.search(pattern, item)
    if match:
        rows_to_check.append(i)
print(
    f"There are {len(df.loc[rows_to_check, :])} remaining entries with digits in info_1 column.\n"
)
print(f'There are {df["age"].isna().sum()} remaining missing values for age.')

There are 0 remaining entries with digits in info_1 column.

There are 218 remaining missing values for age.


<IPython.core.display.Javascript object>

### Observations:
- All of the age data that had been in `info_1` has been successfully extracted.
- There are 218 remaining missing values for `age` that we hope to find in the other info columns.
- We will include the remaining values in `info_1` when extracting citizenship and role information.

In [71]:
df[df["age"].isna()]

day                name  \
6846    17         Spiro Agnew   
12861   14         Muslimgauze   
13511   29    Faustin Birindwa   
14994   15   Željko Ražnatović   
15063   28      Sarah Caudwell   
...     ..                 ...   
130463   3        William Samb   
130578   8        Mun Jong-nam   
130866  19       Paul Kavanagh   
130944  23       Hassan Dhuhul   
131360  10  Desmond Brathwaite   

                                                                                  info  \
6846    , American politician, 77, 39th Vice President of the United States, leukemia.   
12861                           , , 37, British electronic musician, fungal infection.   
13511                                                      , Prime Minister of Zaire .   
14994                                 , , 47, Serbian mobster and paramilitary leader.   
15063                    , , 60, British detective story writer and barrister, cancer.   
...                                                                                ...   
130463                                            , Papua New Guinean politician, MP .   
130578                          , North Korean diplomat, ambassador to Italy , stroke.   
130866                                   , Irish businessman and politician, senator .   
130944                                              , Somali politician, MP , bombing.   
131360                                                 , Saint Lucian politician, MP .   

                                                            link  \
6846                   https://en.wikipedia.org/wiki/Spiro_Agnew   
12861                  https://en.wikipedia.org/wiki/Muslimgauze   
13511             https://en.wikipedia.org/wiki/Faustin_Birindwa   
14994                        https://en.wikipedia.org/wiki/Arkan   
15063               https://en.wikipedia.org/wiki/Sarah_Caudwell   
...                                                          ...   
130463                https://en.wikipedia.org/wiki/William_Samb   
130578                https://en.wikipedia.org/wiki/Mun_Jong-nam   
130866  https://en.wikipedia.org/wiki/Paul_Kavanagh_(politician)   
130944               https://en.wikipedia.org/wiki/Hassan_Dhuhul   
131360          https://en.wikipedia.org/wiki/Desmond_Brathwaite   

       num_references  year      month                            info_1  \
6846              207  1996  September               American politician   
12861              26  1999    January                                     
13511               4  1999      April           Prime Minister of Zaire   
14994              55  2000    January                                     
15063               9  2000    January                                     
...               ...   ...        ...                               ...   
130463              2  2022      March      Papua New Guinean politician   
130578             10  2022      March             North Korean diplomat   
130866              3  2022      March  Irish businessman and politician   
130944              2  2022      March                 Somali politician   
131360              2  2022      April           Saint Lucian politician   

                     info_2                                        info_3  \
6846                     77      39th Vice President of the United States   
12861                    37                   British electronic musician   
13511                   NaN                                           NaN   
14994                    47       Serbian mobster and paramilitary leader   
15063                    60  British detective story writer and barrister   
...                     ...                                           ...   
130463                   MP                                           NaN   
130578  ambassador to Italy                                        stroke   
130866              senator                                           NaN   
130944           

<IPython.core.display.Javascript object>